src/webgpu/shader/validation/parse/shadow_builtins.spec.ts

```diff
@@ -0,0 +1,1002 @@
+  .fn(t => {
+    const data = kTextureTypeTests[t.params.builtin];
+    const value = `let ${data.keyword} = 4;`;
+
+    const module_scope = t.params.inject === 'module' ? `var<private> ${data.keyword} : f32;` : ``;
+    const func = t.params.inject === 'function' ? value : ``;
+
+    const code = `
+${module_scope}
+@group(0) @binding(0) ${data.src}
+
+fn func() {
+  ${func}
+}
+    `;
+    const pass = t.params.inject === 'none' || t.params.inject === 'function';
+    t.expectCompileResult(pass, code);
+  });
+
+const kTextureTests = {
+  textureDimensions: {
+    keyword: `textureDimensions`,
+    src: `_ = textureDimensions(t_2d);`,
+  },
+  textureGather: {
+    keyword: `textureGather`,
+    src: `_ = textureGather(1, t_2d, s, vec2(1, 2));`,
+  },
+  textureGatherCompare: {
+    keyword: `textureGatherCompare`,
+    src: `_ = textureGatherCompare(t_2d_depth, sc, vec2(1, 2), 3);`,
+  },
+  textureLoad: {
+    keyword: `textureLoad`,
+    src: `_ = textureLoad(t_2d, vec2(1, 2), 1);`,
+  },
+  textureNumLayers: {
+    keyword: `textureNumLayers`,
+    src: `_ = textureNumLayers(t_2d_array);`,
+  },
+  textureNumLevels: {
+    keyword: `textureNumLevels`,
+    src: `_ = textureNumLevels(t_2d);`,
+  },
+  textureNumSamples: {
+    keyword: `textureNumSamples`,
+    src: `_ = textureNumSamples(t_2d_ms);`,
+  },
+  textureSample: {
+    keyword: `textureSample`,
+    src: `_ = textureSample(t_2d, s, vec2(1, 2));`,
+  },
+  textureSampleBias: {
+    keyword: `textureSampleBias`,
+    src: `_ = textureSampleBias(t_2d, s, vec2(1, 2), 2);`,
+  },
+  textureSampleCompare: {
+    keyword: `textureSampleCompare`,
+    src: `_ = textureSampleCompare(t_2d_depth, sc, vec2(1, 2), 2);`,
+  },
+  textureSampleCompareLevel: {
+    keyword: `textureSampleCompareLevel`,
+    src: `_ = textureSampleCompareLevel(t_2d_depth, sc, vec2(1, 2), 3, vec2(1, 2));`,
+  },
+  textureSampleGrad: {
+    keyword: `textureSampleGrad`,
+    src: `_ = textureSampleGrad(t_2d, s, vec2(1, 2), vec2(1, 2), vec2(1, 2));`,
+  },
+  textureSampleLevel: {
+    keyword: `textureSampleLevel`,
+    src: `_ = textureSampleLevel(t_2d, s, vec2(1, 2), 3);`,
+  },
+  textureSampleBaseClampToEdge: {
+    keyword: `textureSampleBaseClampToEdge`,
+    src: `_ = textureSampleBaseClampToEdge(t_2d, s, vec2(1, 2));`,
+  },
+};
+
+g.test('shadow_hides_builtin_texture')
+  .desc(`Test that shadows hide texture buitins.`)
+  .params(u =>
+    u
+      .combine('builtin', keysOf(kTextureTests))
+      .combine('inject', ['none', 'function', 'sibling', 'module'])
+      .beginSubcases()
+  )
+  .fn(t => {
+    const data = kTextureTests[t.params.builtin];
+    const value = `let ${data.keyword} = 4;`;
+
+    const module_scope = t.params.inject === 'module' ? `var<private> ${data.keyword} : i32;` : ``;
+    const sibling_func = t.params.inject === 'sibling' ? value : ``;
+    const func = t.params.inject === 'function' ? value : ``;
+
+    const code = `
+@group(0) @binding(0) var t_2d: texture_2d<f32>;
```

Comment:
External textures are always polyfilled. Probably worth testing that too.

Reply:
I'm not sure what this means, what are you looking to see tested?  These declarations exist in order to be used by the kTextureTests calls above.

Do you want one of those to reference an external texture?